### ISSUE TITLE
Refactor `drasil-gool`

### ISSUE THREAD

**Issue by B-rando1** [JacquesCarette/Drasil#3934]
The changes I made this summer went pretty well, but the organization of `drasil-gool` fell behind a bit.  Some of the structure and naming is a bit misleading.
- It is called `drasil-gool` when it contains both GOOL and GProc.
- All of its modules are in `drasil-gool/Drasil/GOOL`.
- Some of the helper files have misleading names and/or have functions in the wrong places:
  - `LanguagePolymorphic.hs` has functions that the Julia renderer doesn't use.
  - `CommonPseudoOO.hs` has functions that are Shared, i.e. not intrinsically OO.
  - There may be others.  For example, `CLike.hs` may have functions that are used by languages that are not particularly C-like, e.g. Julia.
- There are some naming inconsistencies in both `drasil-gool` and `drasil-code`:
  - Procedural is sometimes referred to as `Proc`, sometimes as `GProc`.
  - Features common to both generic languages are referred to as `Common` or `Shared`.
- There's a lot of code duplication between `CodeInfoOO.hs` and `CodeInfoProc.hs`.

I would recommend making the following changes:
- [ ] I don't think it's overly important to rename `drasil-gool`, but if we can find a way to refer to GOOL and GProc collectively, that could be nice.  But we would need to use the name consistently for it to be a positive change.
  - `drasil-genlang` sounds good, but `drasil-gen` and `drasil-lang` exist 😄.
- [ ] I think we should split up `drasil-gool/Drasil` into `GOOL`, `GProc`, and `Shared` folders.
- [ ] We should inventory the helper functions (`LanguagePolymorphic.hs`, `CommonPseudoOO.hs`, etc.).  We'll want to rename files and move functions to different files, so that the name of each file accurately describes the kinds of functions that can be found in it.
- [ ] We need to pick a name for procedural and shared programs and stick with it.  I like `GProc` and `Shared` personally.
- [ ] We should merge the implementations for most of the typeclass methods in the `CodeInfoX.hs` files.  This should be pretty easy, and will look a lot like the `LanguagePolymorphic.hs` functions for the renderers.

**Comment by balacij** [comment]
@sarrasoussia Do you want to try having a look at some of the points @B-rando1 mentioned? I was thinking specifically the last point might be a good starting point.
>  We should merge the implementations for most of the typeclass methods in the CodeInfoX.hs files. This should be pretty easy, and will look a lot like the LanguagePolymorphic.hs functions for the renderers.

**Comment by B-rando1** [comment]
I appreciate anything you can do here, it will certainly make my life easier in the fall 🙂 

Regarding my last point above, I agree that it could be a good place to start. I believe [this](https://github.com/JacquesCarette/Drasil/pull/3863/files#diff-296d2e1f7849ab94bafc54fa702cdc14b9c2f070fc73426cd94667684b412327) is the PR I was referring to. There's a lot of changes, but if you zoom in on the `LanguagePolymorphic` diff it should give you the right idea.

I can't quite remember if I was envisioning us ending up with 3 files (`CodeInfoShared`, `CodeInfoProc`, `CodeInfoOO`) or 1 containing everything. The `RendererClassesX` files follow the former, while the gooltest (e.g. [here](https://github.com/JacquesCarette/Drasil/blob/main/code/drasil-code/test/HelloWorld.hs)) files consolidate everything into a single file. Either should work, and @balacij and @JacquesCarette should be able to help you make the decision.

Hope this helps!

**Comment by JacquesCarette** [comment]
`drasil-codegen` is probably best, at least for our current level of understanding.

**Comment by sarrasoussia** [comment]
I dont understand why this point is an issue: LanguagePolymorphic.hs has functions that the Julia renderer doesn't use.

If it's a shared toolbox, then Julia just doesn’t need all the tools in it. Does the unused code impacts performance?

**Comment by sarrasoussia** [comment]
Julia does use LanguagePolymorphic, but only a subset of its functions, similarly for the other renderers so i think no chqnges should be made to `LanuagePolymorphic.hs`

<img width="583" alt="Image" src="https://github.com/user-attachments/assets/8b303515-27d2-4dcb-a1dc-c99e3d4f0fc3" />

**Comment by sarrasoussia** [comment]
I checked the `CodeInfoProc.hs` and `CodeInfoOO.hs` and found some common instances:

- SharedProg
- ProgramSym
- FileSym
- BodySym
- BlockSym
- TypeSym, TypeElim
- ScopeSym
- VariableSym, VariableElim
- ValueSym
- Argument, Literal, MathConstant, VariableValue, CommandLineArgs
- NumericExpression, BooleanExpression, Comparison, ValueExpression
- FunctionSym
- List, Set, InternalList
- ThunkSym, ThunkAssign
- VectorType, VectorDecl, VectorThunk, VectorExpression
- StatementSym, AssignStatement, DeclStatement
- IOStatement, StringStatement
- FuncAppStatement, CommentStatement, ControlStatement
- VisibilitySym, ParameterSym, MethodSym
- ModuleSym

Im thinking of refactoring all the shared instance implementations into a common base module `Drasil.Shared.CodeInfoBase
`

**Comment by B-rando1** [comment]
@balacij asked me to see if this issue is still relevant. Most of the big items mentioned are complete, but there are a few things we haven't done yet:
- The module is called `drasil-gool` when it contains both GOOL and GProc. There was some discussion of renaming it to `drasil-codegen` - is that still something we want to do?
- Some of the helper files like LanguagePolymorphic might still contain functions that don't match the file name. This isn't a huge problem, but it might be good for me to do a pass over these files and see if there's anything we can move around to make things easier to find.
- Lots of code duplication between `CodeInfoOO` and `CodeInfoProc`. Afaik this is still true. I'm not sure if there's an easy way to merge their common implementations, but I feel like there should be.

**Comment by JacquesCarette** [comment]
- `drasil-codegen` is not bad. But I wouldn't want to rename until we've done some brainstorming.
- small step refactors are always good, especially those that increase explainability
- I would think that pulling out a `CodeInfoGeneric` that contains the commonalities, and then have both `CodeInfoOO` and `CodeInfoProc` contain a `CodeInforGeneric` would be a good way to deal with that.

**Comment by B-rando1** [comment]
I agree we should think it through before committing to `drasil-codegen`. Codegen is more accurate than GOOL now that we have procedural languages, but  GOOL is still more fun to say 😀
